Uptodate Source pointing to GitHub in RPM spec

pcahyna opened the pull request:
##### Pull Request Details:

* Type: **Bug Fix**

* Impact: **Low**

* Reference to related issue (URL):  closes #2945

* How was this pull request tested?
`spectool -g packaging/rpm/rear.spec` downloads the correct tarball

* Brief description of the changes in this pull request:
Change the URL in the Source tag in the RPM spec file to point to the tarball on GitHub instead of SourceForge (where the tarball is not updated anymore).
AFAIK GitHub creates the tarballs automatically from tags.

pcahyna commented:
ToDo: the Makefile also needs updating, otherwise it will rewrite the spec when executing`make dist`.

schlomo commented on packaging/rpm/rear.spec:
```diff
@@ -18,8 +18,7 @@ License: GPL-3.0
 
-# as GitHub stopped with download section we need to go back to Sourceforge for downloads
-Source: https://sourceforge.net/projects/rear/files/rear/%{version}/rear-%{version}.tar.gz
+Source: https://github.com/rear/rear/archive/%{version}.tar.gz#/rear-%{version}.tar.gz
```
Nice, didn't know this trick. Do we need the `#...` suffix? I tried several URLs like

* https://github.com/rear/rear/archive/galaxy11.tar.gz
* https://github.com/rear/rear/archive/531d9074d584ba4192da51ee5b7ab0fad460d57b.tar.gz
* https://github.com/rear/rear/archive/2.7.tar.gz
* https://github.com/rear/rear/archive/master.tar.gz

And they all yielded archives with a matching `rear-<tag/branch/commit>` names *and* subdirectories

pcahyna replied:
This trick is documented here: https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_troublesome_urls , with examples here: https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_git_tags .
In general, it is needed so that the tarballs are named `%{project}-%{version}.tar.gz` and not just `%{version}.tar.gz` in the case of a typical archive URL like `https://github.com/rear/rear/archive/%{version}.tar.gz`.
I don't think you get a `rear-<tag/branch/commit>`  name with such an URL, only `<tag/branch/commit>` (`.tar.gz`). See:
```
wget https://github.com/rear/rear/archive/2.7.tar.gz
--2023-03-01 12:47:12--  https://github.com/rear/rear/archive/2.7.tar.gz
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://codeload.github.com/rear/rear/tar.gz/refs/tags/2.7 [following]
--2023-03-01 12:47:13--  https://codeload.github.com/rear/rear/tar.gz/refs/tags/2.7
2023-03-01 12:47:14 (1.45 MB/s) - ‘2.7.tar.gz’ saved [852253]
```
And note that this URL is not used just to know where to download the file from. `rpmbuild` extracts its last component (basename) to know what is the local archive name, even if the file is pre-downloaded. `2.7.tar.gz` is not correct.
In the case of ReaR, the trick is unnecessary, because there have always been `rear-%{version}` tags in addition to the bare `%{version}` tags in the repo (this is not typical for most projects/repositories). Thanks to this, I can simply specify `Source:` as `https://github.com/rear/rear/archive/%{name}-%{version}.tar.gz` and get rid of the trick.

schlomo replied:
Ah, I learned something new: `curl` and `wget` ignore - by default - the `content-disposition: attachment; filename=rear-2.7.tar.gz` header that GitHub actually sets.

`curl -OJ` actually works:
<img width="849" alt="image" src="https://user-images.githubusercontent.com/101384/222220220-48592ad9-4689-4092-96ad-a93323f09736.png">

Good that you found a way around that problem

pcahyna replied:
Ah, I learned something new as well, I did not know that there is such a header (I assumed that wget would use it if it were set).
Anyway, respecting it would not have helped, because the file would get downloaded under the correct name, but rpmbuild would not be able to figure out what the name is (as it extracts the last component of the URL to get the file name).

pcahyna replied:
> In the case of ReaR, the trick is unnecessary, because there have always been `rear-%{version}` tags in addition to the bare `%{version}` tags in the repo (this is not typical for most projects/repositories). Thanks to this, I can simply specify `Source:` as `https://github.com/rear/rear/archive/%{name}-%{version}.tar.gz` and get rid of the trick.

It turns out that I can not, because https://github.com/rear/rear/archive/rear-2.7.tar.gz is a text file that says only
```
the given path has multiple possibilities: #<Git::Ref:0x00007f4e90a15960>, #<Git::Ref:0x00007f4e90a14f60>
```
This is probably because there is both a branch named [rear-2.7](https://github.com/rear/rear/tree/rear-2.7) and a tag named [rear-2.7](https://github.com/rear/rear/releases/tag/rear-2.7), and similarly for the older releases (2.6, 2.5). So I have to go back to the original idea of using bare `%{version}`, in the archive name and keep the trick, like 2.7.tar.gz.

In the future it would probably be a good idea to use a release branch name that does not shadow the release tag name, like e.g. `release-2.8`.

jsmeix replied:
@pcahyna 
I did the ReaR 2.8 release (my first release ever) according to
https://github.com/rear/rear/wiki/Release-process
as far as I could understand the instructions there and
partially in some kind of "bona fide script kiddie" mode ;-)
so feel free to enhance
https://github.com/rear/rear/wiki/Release-process
as needed.

schlomo reviewed (approved):
I think you can simply remove the corresponding `sed` line from the `Makefile` and I would appreciate for you to do that as well within this PR.

About Gentoo: According to https://devmanual.gentoo.org/ebuild-writing/variables/index.html#renaming-sources I'd expect the URL line to look like this:

```
SRC_URI="https://github.com/rear/rear/archive/${PV}.tar.gz -> ${P}.tar.gz"
```

But indeed, a test by somebody using Gentoo would be nice, although I'd rather make this change blindly as not to change it - the old URL is 100% broken, the new one might just work.

DEB and Arch don't carry the dist archive URL in the description so that we don't need to change it there.

jsmeix reviewed (approved):
I blindly approve it because I trust you and
I know almost nothing about ReaR's make and build code.

For SUSE and openSUSE I use my own RPM spec files
to be in compliance with what SUSE requires
(certain SUSE specific rules).

pcahyna commented:
> I think you can simply remove the corresponding `sed` line from the `Makefile` and I would appreciate for you to do that as well within this PR.
> 
> About Gentoo: According to https://devmanual.gentoo.org/ebuild-writing/variables/index.html#renaming-sources I'd expect the URL line to look like this:
> 
> ```
> SRC_URI="https://github.com/rear/rear/archive/${PV}.tar.gz -> ${P}.tar.gz"
> ```
> 
> But indeed, a test by somebody using Gentoo would be nice, although I'd rather make this change blindly as not to change it - the old URL is 100% broken, the new one might just work.

ok, will do

schlomo commented:
Yes, agree with you that a more explicitly named tag would be useful.

pcahyna commented:
> I think you can simply remove the corresponding `sed` line from the `Makefile` and I would appreciate for you to do that as well within this PR.

This won't work properly, because `make dist` constructs a tarball named e.g. ` rear-2.7-git.4965.531d9074.githubsourceinspecfile.changed.tar.gz` under `dist`, so the `Source:` line needs to be updated to point to this tarball, otherwise `rpmbuild` invoked from `make srpm` would not find it (just tried that). But I can update the `sed` line to use a working URL (i.e. GitHub) for `Source`.

pcahyna commented:
> ToDo: the Makefile also needs updating, otherwise it will rewrite the spec when executing`make dist`.

it won't, because Makefile does not touch `packaging/rpm/rear.spec`, it constructs a temporary spec under `dist` (the message `Rewriting packaging/rpm/rear.spec, packaging/debian/rear.dsc and usr/sbin/rear` is misleading).

pcahyna commented on packaging/rpm/rear.spec:
```diff
@@ -18,8 +18,7 @@ License: GPL-3.0
-Source: https://sourceforge.net/projects/rear/files/rear/%{version}/rear-%{version}.tar.gz
+Source: https://github.com/rear/rear/archive/%{version}.tar.gz#/rear-%{version}.tar.gz
 
 # BuildRoot: is required for SLES 11 and RHEL/CentOS 5 builds on openSUSE Build Service (#2135)
```
@jsmeix please remind me - is SLES 11 going to be deprecated in the next release? If so, we could remove this line.

jsmeix replied:
SLES11 is no longer officially supported by ReaR upstream
since a long time and since ReaR 2.7 is released even I
do no longer care about SLE11 at ReaR upstream
so feel free to remove such old and obsolete stuff.

Explanation:
During ReaR 2.7 development I only liked to
not knowingly break ReaR 2.7 for SLE11 because
I might need ReaR 2.7 for SLE11 for special SUSE customers
on very specific request and therefore I liked to avoid
that I would have to fix various things at various places
to make ReaR 2.7 usable for a specific use case on SLE11.

schlomo commented:
Good catch. I was also fighting our `Makefile` yesterday and would love to find a way how ReaR can run from source without modifying the source area at all.

For the `Source:` entry in the spec file, is any valid URL OK or must it actually match the content and the dist archive file name? I find it very important that the dist archive file name can continue to carry the full git branch and version info.

pcahyna commented:
> For the `Source:` entry in the spec file, is any valid URL OK or must it actually match the content and the dist archive file name? I find it very important that the dist archive file name can continue to carry the full git branch and version info.

It does not need to be a URL at all. I would actually prefer the Makefile to keep only the file name in Source. URLs of type `https://sourceforge.net/projects/rear/files/rear/2.7/rear-2.7-git.4965.531d9074.githubsourceinspecfile.changed.tar.gz`, which the current Makefile produces, have never been correct even when SourceForge download was working and won't became more correct by changing them to github.com. I would prefer just `Source: rear-2.7-git.4965.531d9074.githubsourceinspecfile.changed.tar.gz` which expresses clearly that the tarball has been built locally and can't be found anywhere on the Internet.

pcahyna commented:
> Good catch. I was also fighting our `Makefile` yesterday and would love to find a way how ReaR can run from source without modifying the source area at all.

have a look at the `dist` target - it produces modified sources under `build` without touching the original sources (except `man`).

pcahyna commented:
by the way, I believe ReaR can already run from source checkout without modifications - https://github.com/rear/rear#quick-start-guide

schlomo commented:
The problem is that ReaR running from source add os.conf and the log files
into the source tree. And uses it to build the rescue image.

And yes, running make shouldn't randomly recreate the manpages and docs
with a different asciidoc heading embedded.

pcahyna commented:
> For the Source: entry in the spec file, is any valid URL OK or must it actually match the content and the dist archive file name? 

if by `dist archive file name` you mean the archive file name that `rpmbuild` will actually use to build the source RPM, the last component of the `Source` URL must match it. This is because the last component of the URL is THE way to determine the source tarball name that rpmbuild uses. There is no independent way to provide it. OTOH, the URL does not need to be valid and indeed, as I wrote above, it does not need to be an URL at all, a local file name is enough.

pcahyna commented:
> The problem is that ReaR running from source add os.conf and the log files into the source tree. And uses it to build the rescue image.

So the problem is that the rescue image will contain `os.conf` and log files? But IIRC this will be the case also for installed ReaR.

pcahyna commented:
@schlomo I think I addressed all the ToDos now.

schlomo commented on Makefile:
```diff
@@ -163,15 +163,17 @@ uninstall:
+	@echo -e "\033[1m== Rewriting build/$(name)-$(distversion)/{$(specfile),$(dscfile),$(rearbin)} ==\033[0;0m"
 	sed -i.orig \
-		-e 's#^Source:.*#Source: https://sourceforge.net/projects/rear/files/rear/${version}/$(name)-${distversion}.tar.gz#' \
+		-e 's#^Source:.*#Source: $(name)-${distversion}.tar.gz#' \
```
So this `sed` patches just the dist archive name into the `Source` field of the spec file, doesn't that make it completely irrelevant what is says in the actual spec file below, as every `make rpm` includes also a `make dist` before?

pcahyna replied:
I think it is good to document where the canonical source can be obtained from.

schlomo commented on packaging/rpm/rear.spec:
```diff
@@ -18,8 +18,7 @@ License: GPL-3.0
 
-# as GitHub stopped with download section we need to go back to Sourceforge for downloads
-Source: https://sourceforge.net/projects/rear/files/rear/%{version}/rear-%{version}.tar.gz
+Source: https://github.com/rear/rear/archive/%{version}.tar.gz#/rear-%{version}.tar.gz
```
In which circumstance or context does this line actually take effect if `make dist` always changes that to just the dist archive file name?

pcahyna replied:
Only when you want to build the RPM manually and need to know where to obtain the sources from. In the future, we might also avoid spec file rewriting in `make dist` if `OFFICIAL` is set and thus the RPM is goping to be built from unmodified sources.

schlomo reviewed (approved):
Looks good, not sure about the context where the full URL in the spec file will be used though.

For sure I see no harm so if it helps anybody then this is great.

pcahyna commented:
> Looks good, not sure about the context where the full URL in the spec file will be used though.
> 
> For sure I see no harm so if it helps anybody then this is great.

The usual RPM build automation (`rpmbuild` and `koji`) do not use the full URL, only the last component (at least in Fedora, maybe the tooling for other distros is actually using it). The full URL is there mostly to document where the source can be obtained. Note that https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_troublesome_urls mentions cases where a URL can't be used, so only the file name is used in `Source` and the full URL is in a comment.

There is a tool though called `spectool` that automates the download of the source file : so instead of downloading the tarball manually, you can execute `spectool -g packaging/rpm/rear.spec`. For this, it is handy to have a correct URL. There may be other tooling that needs valid URL. (Packit builds executed from PRs do not need it though, because Packit rebuilds the tarball including the changes and changes `Source` to point to the tarball that it has built, in analogy with `make dist`.)

All in all, this PR is less important than it may seem. The incorrect URL was not causing any failure.

schlomo commented:
I'm also not super happy with out Makefile, but this PR is a good improvement and we learned something. Thanks a lot!

jsmeix commented:
In the openSUSE Build Service projects
Archiving:Backup:Rear and Archiving:Backup:Rear:Snapshot
I removed the build repository "SLE_11_SP4"
so ReaR gets no longer built there for SLE11, cf.
https://github.com/rear/rear/pull/2949#discussion_r1124225289
and old stuff for SLE11 in packaging/rpm/rear.spec
can be safely removed.

pcahyna commented:
> @pcahyna I did the ReaR 2.7 release (my first release ever) according to https://github.com/rear/rear/wiki/Release-process as far as I could understand the instructions there and partially in some kind of "bona fide script kiddie" mode ;-) so feel free to enhance https://github.com/rear/rear/wiki/Release-process as needed.

@jsmeix thank you for showing us how this Makefile target is actually being used in the release process. If we decide to change the functionality of the target, we should definitely consider how it fits into the release process.

pcahyna commented:
@schlomo 
> Yes, agree with you that a more explicitly named tag would be useful.

(this was in reaction to my

> there is both a branch named [rear-2.7](https://github.com/rear/rear/tree/rear-2.7) and a tag named [rear-2.7](https://github.com/rear/rear/releases/tag/rear-2.7), and similarly for the older releases (2.6, 2.5). So I have to go back to the original idea of using bare %{version}, in the archive name and keep the trick, like 2.7.tar.gz.
> 
> In the future it would probably be a good idea to use a release branch name that does not shadow the release tag name, like e.g. release-2.8.

)

I think that the tag name is actually OK and one should use a different release branch name to avoid ambiguity.
